### ISSUE TITLE
restjavad restarts with 3.19.0

### ISSUE THREAD

**Issue by wncocz** [F5Networks/f5-appsvcs-extension#263]
### Environment
 * Application Services Version: 3.19.0
 * BIG-IP Version: 13.1.3

### Summary

We have been seeing many failed declaration deployments due to restjavad restarts since moving from AS3 3.17.1 to 3.19.0.

At the same time, we have transitioned from using the bigip_appsvcs_extension Ansible module from the f5devcentral.f5ansible galaxy role to using the f5devcentral.atc_deploy galaxy role.  With regard to the actual AS3 POST step, the difference there is that we've moved from sync to async deployments.  BUT once we see the failures happening with atc_deploy, we see the same behavior with bigip_appsvcs_extension.

### Steps To Reproduce

I don't know if there is a clear/consistent way to reproduce this problem.  For us, we see it when POSTing a "typical" virtual + pool + profiles declaration, sometimes with a single HTTPS virtual + pool + profiles + cert & key.

I don't know if it only happens when POSTing an app+tenant that already exists (changing a monitor, for example), but that is where I have noticed it.  I have not experimented to see if it happens when posting a previously-nonexistant tenant.

### Expected Behavior

Successful deployment

### Actual Behavior

Deployment fails.  Digging into logs on the BIGIP, we find this:

In /var/log/icrd, a whole lot of these:
```
Apr 24 15:41:10 slot3/HOSTNAME notice icrd_child: 20698,20720,    iControl REST Child Daemon,WARNING,Ignoring unexpected json property at configuration item /ltm/data-group/internal/~Common~appsvcs~dataStore/records/as3_async_records##. The json property is "type".
```
In restjavad.0.log:
```
[SEVERE][138][24 Apr 2020 15:41:11 UTC][DieOnUncaughtErrorHandler] Restjavad exiting immediately due to an uncaught Error
[SEVERE][139][24 Apr 2020 15:41:11 UTC][DieOnUncaughtErrorHandler] Uncaught Error causing restjavad to exit. Thread id 15, 'Non-Blocking threadPool_5': java.lang.OutOfMemoryError: Java heap space
        at java.util.Arrays.copyOfRange(Arrays.java:2694)
        at java.lang.String.<init>(String.java:203)
        at java.nio.HeapCharBuffer.toString(HeapCharBuffer.java:561)
        at java.nio.CharBuffer.toString(CharBuffer.java:1205)
        at com.f5.rest.common.HttpSocketHelper$2.run(HttpSocketHelper.java:367)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:473)
        at java.util.concurrent.FutureTask.run(FutureTask.java:262)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:178)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:292)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1152)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:622)
        at java.lang.Thread.run(Thread.java:748)
```

Leading up to this, as we tail -f the audit log, we can see the dg /Common/appsvcs/dataStore getting loaded up with huge amounts of data.  Or it seems huge to me compared to what I saw using sync on 3.17.1.

So, I tried delete ltm data-group internal /Common/appsvcs/dataStore

Re-ran my deployment, and it succeeded.
This is our "known workaround" now -- delete the dg /Common/appsvcs/dataStore just before the declaration gets POSTed.

**Comment by dstokesf5** [comment]
It looks like restjavad might be having some memory problems, could you try the following steps on your BIG-IP to see if it helps with this issue?
```
Increase memory assigned to the Linux host:

# tmsh modify sys db provision.extramb value 1000 # (value dependent on platform) -- Need to get the correct value for Google platform

2) Allow restjavad to access the extra memory:

# tmsh modify sys db restjavad.useextramb value true

3) Save the config:

# tmsh save sys config

4) The re-provisioning will trigger a restart of the services. Wait until the unit is online again.

5) Increase the restjavad maxMessageBodySize property:

# curl -s -f -u admin: -H "Content-Type: application/json" -d '{"maxMessageBodySize":134217728}' -X POST http://localhost:8100/mgmt/shared/server/messaging/settings/8100 | jq .
```

**Comment by wncocz** [comment]
> It looks like restjavad might be having some memory problems, could you try the following steps on your BIG-IP to see if it helps with this issue?
> 
> ```
> Increase memory assigned to the Linux host:
> 
> # tmsh modify sys db provision.extramb value 1000 # (value dependent on platform) -- Need to get the correct value for Google platform
> 
> 2) Allow restjavad to access the extra memory:
> 
> # tmsh modify sys db restjavad.useextramb value true
> 
> 3) Save the config:
> 
> # tmsh save sys config
> 
> 4) The re-provisioning will trigger a restart of the services. Wait until the unit is online again.
> 
> 5) Increase the restjavad maxMessageBodySize property:
> 
> # curl -s -f -u admin: -H "Content-Type: application/json" -d '{"maxMessageBodySize":134217728}' -X POST http://localhost:8100/mgmt/shared/server/messaging/settings/8100 | jq .
> ```

While this may work around the issue, I'm not confident it is the right solution for us.  Doing this to every VCMP guest we have deployed will require quite a lot of scheduling & red tape.

Again, this is new behavior from 3.17.1 to 3.19.0.  And the declaration I'm posting can be a pretty small -- in one case I whittled it down to a one-connect profile and a tcp-half-open monitor in /Common/Shared.

And "delete ltm data-group internal /Common/appsvsc/dataStore" is much less intrusive than reconfiguring the memory provisioning of the instances.  The fact that it works leads me to believe that AS3 should be able to resolve this within AS3.  But I don't know what all that data I'm deleting is.  I will see what else I can find out next time it occurs.

**Comment by dstokesf5** [comment]
Are you seeing the same behavior with 3.18?

**Comment by erjac77** [comment]
We have also encountered this problem since the update from 3.13.2 to 3.19.0. Fortunately for us, we have only run 3.19.0 in our laboratory so far. Restarting 'restjavad' is not a problem for us, but it will surely be when we go into production...

**Comment by dstokesf5** [comment]
We are continuing to look into this. It is being tracked in our internal product backlog as AUTOTOOL-1524.

**Comment by wncocz** [comment]
> Are you seeing the same behavior with 3.18?

When I encounter it again, I will try a downgrade to 3.18.0 and see what happens.

**Comment by dstokesf5** [comment]
The data store is storing task records. Before 3.19, those task records were only from asynchronous requests. Synchronous requests were added to these records in 3.19 to simplify the code base and allow for more features in the future, and is most likely why you are seeing an increase in the data store size. As far as I can tell we have not been seeing restjavad restarts in our testing environments, which get hit with many AS3 requests every night. I am wondering if we are not seeing this because of the environment we are running AS3 in.

How many CPUs does your BIG-IP have?
How much memory does your BIG-IP have?
You mentioned the BIG-IP version was 13.1.3. Is that 13.1.3.0, or a different fourth element release?

**Comment by wncocz** [comment]
> The data store is storing task records. Before 3.19, those task records were only from asynchronous requests. Synchronous requests were added to these records in 3.19 to simplify the code base and allow for more features in the future, and is most likely why you are seeing an increase in the data store size. As far as I can tell we have not been seeing restjavad restarts in our testing environments, which get hit with many AS3 requests every night. I am wondering if we are not seeing this because of the environment we are running AS3 in.
> 
> How many CPUs does your BIG-IP have?
> How much memory does your BIG-IP have?
> You mentioned the BIG-IP version was 13.1.3. Is that 13.1.3.0, or a different fourth element release?

It's also possible that we are seeing dataStore getting huge because we just started using async requests at the same time we upgraded to 3.19.0 (and started using atc_deploy instead of bigip_appsvcs_extension).

The instances on which we're seeing this are mostly (all?) VCMP guests on C2400 chassis w/B2250 blades.  Guests are limited to a single blade each, up to 8 cores per blade.  The main offender (the one we make the most changes to) is running on 8 cores.  The guest's (and host's for that matter) software is specifically 13.1.3 ENG HF 0.21.6.

**Comment by wncocz** [comment]
I have a little bit more information.  First, this continues on appsvcs 3.19.1.

Second, our dataStore data-groups can get quite large without even POSTing any declaration to the BIG-IP.  I have an Ansible playbook that loops over a set of applications doing the following:
1. (local Ansible) Parse the app template into AS3 declaration
2. (AS3 GET) GET the most recent declaration from BIG-IPs it was (possibly) POSTed to
3. (non-AS3) Gets virtual/pool facts from the BIG-IPs and attempts to build an in-memory declaration that represents actual running config on the BIG-IP.
4. (local Ansible) Compares the parsed app template with the results of (2) & (3), and spews the diff result.

We use this both for finding changes made to BIG-IPs outside of AS3 and for regression-testing our app template framework.

None of these steps actually sends a declaration to the BIG-IP, but somewhere in there, the dataStore is growing.  In some cases, quite large.  After running this on a set of applications and noting that one unit experienced several restjavad restarts during the test, I ran "tmsh -a list ltm data-group internal /Common/appsvcs/dataStore | wc -c" on 25 BIG-IPs, and I got numbers ranging from 72272 up to 8775376.  The unit whose restjavad restarted several times was the 8775376.

**Comment by dstokesf5** [comment]
@wncocz Would it be possible to get a qkview of the device with the affected data group? I would love to be able to inspect it and get a better idea of what is ending up in there. So far our investigations have not found anywhere that unexpected data should be ending up in the data groups. However, the data group growing that large with GET requests certainly sounds odd.

**Comment by wncocz** [comment]
> @wncocz Would it be possible to get a qkview of the device with the affected data group? I would love to be able to inspect it and get a better idea of what is ending up in there. So far our investigations have not found anywhere that unexpected data should be ending up in the data groups. However, the data group growing that large with GET requests certainly sounds odd.

F5 Case # C3275936.  Uploading qkview to iHealth now.

**Comment by dstokesf5** [comment]
I was finally able to get access to the data I needed from the qkview. I do not see any unexpected data in the dataStore data group. The GET requests also makes sense since these still end up in the command history (which i believe is limited to 25 items). I suspect you would see the same behavior with asynchronous requests on previous versions of AS3.

It looks like some data group updates can get large enough to overwhelm restajavad. We are looking into ways to reduce this load on restjavad.

**Comment by wncocz** [comment]
> I was finally able to get access to the data I needed from the qkview. I do not see any unexpected data in the dataStore data group. The GET requests also makes sense since these still end up in the command history (which i believe is limited to 25 items). I suspect you would see the same behavior with asynchronous requests on previous versions of AS3.
> 
> It looks like some data group updates can get large enough to overwhelm restajavad. We are looking into ways to reduce this load on restjavad.

Yes, it happens frequently for us.  Is there anything we can do to reduce the amount of data sent to the dataStore?  If there's something we can do differently such that we don't trigger these restjavad crashes as frequently, that would be great.

**Comment by dstokesf5** [comment]
@wncocz We are still looking at ways to mitigate this issue. I do have one possible way for you to reduce the data sent to the data store, but it is not pretty.

You could try modifying `/var/config/rest/iapps/f5-appsvcs/lib/asyncHandler.js` to reduce the number of total records stored:
```javascript
const MAX_RECORDS = 25;
```
And restarting restnoded with `bigstart restart restnoded`.

If you are not interested in the history of async records, you should be able to safely set this down to  5.

**Comment by dstokesf5** [comment]
Do you see a significant reduction in the size of the data store if you do a GET request to the `mgmt/shared/appsvcs/task` endpoint?

**Comment by wncocz** [comment]
@dstokesf5 I prefer to avoid making a change to the AS3 code across all these devices.

Is there a purpose for storing these tasks other than being able to query their status?

Can we get support for DELETE /mgmt/shared/appsvcs/task/{uuid} ?  (I tried it, but I get a java.util.concurrent.TimeoutException, and the task does not get deleted from the list.)

We have no need to have tasks stored once they are complete.  getAsyncResponse() comments say they stick around for 24 hrs, but the EXPIRE_TIME is actually a week.  We're done with them within the minute, usually.

atc_deploy ansible role could be updated to delete completed tasks, and/or I could write a simple playbook to purge tasks from our devices periodically (or after a major deployment is complete).

**Comment by dstokesf5** [comment]
@wncocz I suspect one issue you are running into is not giving AS3 an opportunity to trigger it's task clean up. That is why I asked the question about doing a GET request to the `mgmt/shared/appsvcs/task` endpoint. Due to some original design decisions, querying the task endpoint is the only time AS3 is able to clean up its task records. We are looking into fixing that, but I want to be sure that fix will also help your case.

**Comment by wncocz** [comment]
@dstokesf5 Ah yes, I see that now.  And yes, I can reduce the size (a bit) by doing a GET.  I suppose that depends on how old the tasks are; I recall doing a large deployment late last week that is probably sticking around for now.

I also see where it says:

     * If the request was completed, the record will
     * also be deleted.

But I do not see where that is actually happening.

On one unit, I changed EXPIRE_TIME to ```const EXPIRE_TIME = 60 * 60 * 1000; // 1 hour```

I restarted restnoded, then ```GET .../task```, and it returned zero results.  But dataStore was still large (```list ltm dat.... | wc -c``` => 4106392).  I deleted the data-group, re-ran the GET, and then dataStore was re-created with a single small record.

**Comment by wncocz** [comment]
Also, I forgot to mention: I am using the atc_deploy role for deployment.  It POSTs the declaration, the continuously GETs status until it's complete or until it fails (or until restnoded crashes/restarts) -- these GETs should give AS3 the opportunity to do its cleanup.  Right?

**Comment by dstokesf5** [comment]
We changed over time the clean up rules to be time and count based. So that piece of information is no longer accurate. Is atc_deploy doing GET requests to the task endpoint, or just to the declare endpoint?

**Comment by dstokesf5** [comment]
Just saw your last bit about the GET results. That is odd. I believe that data store should be empty if no tasks were returned.

**Comment by wncocz** [comment]
```
- name: Wait for AS3 Task to complete
  uri:
    url: "https://{{ provider.server }}:{{ provider.server_port }}/mgmt/shared/appsvcs/task/{{ atc_AS3_result.json.id }}"
    method: GET
```

**Comment by dstokesf5** [comment]
Thanks. It looks like there is an even deeper issue here.

**Comment by wncocz** [comment]
Any progress?  F5 Support closed my case on me.

On one of the units that gets more frequent configuration changes, I set EXPIRE_TIME to 10min on June 8 and cleared the dataStore.  Today I noticed that restnoded keeps restarting.  dataStore has records up to as3_async_records139 and "list ltm data-group ... | wc -c" is over 9million.

**Comment by dstokesf5** [comment]
Developers are looking into this during our current sprint.

**Comment by dstokesf5** [comment]
I would also like to apologize for the delays in getting this resolved. This has been difficult to root cause with multiple false leads.

**Comment by dstokesf5** [comment]
@wncocz Still looking into this, but could use some additional information. Is the device that you are seeing these problem on part of an HA cluster?

**Comment by wncocz** [comment]
> @wncocz Still looking into this, but could use some additional information. Is the device that you are seeing these problem on part of an HA cluster?

Yes, all our LTM devices are deployed in a/b pairs.

**Comment by dstokesf5** [comment]
This issue has been root caused to dataStore data sometimes appending instead of overwriting data. The fix for this will be available in the upcoming 3.21 release.

**Comment by wncocz** [comment]
> This issue has been root caused to dataStore data sometimes appending instead of overwriting data. The fix for this will be available in the upcoming 3.21 release.

We started deploying 3.21.0 today.  So far, so good.  Will run an enterprise regression test and let you know how it looks.

**Comment by m-kratochvil** [comment]
Sorry for hijacking the discussion, but after reading this whole thread I see we have the same symptoms in our environment as @wncocz describes:
- java and restnoded processes oom killed frequently
- java process takes up 2.5GB VIRT memory
- cpu at 100%, to the point tmm crashes, device reboots
- massive `bigip.conf` - 13MB, holding all the AS3 internal data-groups
- extensive audit logging, the `/var/log` partition is filling up to 100% twice a day
- the `/var/log/messages` file flooded with strange short hashes like `dSq6Ddw5bUwSg5DBB+QHvBdGrRMeEZ9`

So far I moved 2.5GB of memory to the linux host, disabled audit logging completely but it helped just a little bit
Every config sync operation gets the CPU to 100%

The systems are very unstable after moving on AS3 from the old LBaaSv2. We're running our Big-IPs in a private cloud environment with frequent configuration changes but I didn't forsee such a performace degradation and resource issues that will come with AS3. We are also running vcmp with HA pairs, like @wncocz 

I have several cases opened with F5 support, so far nothing conclusive..

**Comment by wncocz** [comment]
So far, my experience with 3.21 is terrific.  My "big test" succeeded with zero restjavad restarts.

In case this is useful to anyone, here's what I had been doing periodically to clean things up:

# clear dataStore DG
ssh $host tmsh -a delete ltm data-group internal /Common/appsvcs/dataStore
# grab a login token
token=$(curl -sk -d '{"username":"adminuser","password":"adminpass","loginProviderName":"tmos"}' https://$host/mgmt/shared/authn/login 2>/dev/null | jq .token.token -r)
# fetch the task list; this query triggers a task expiration
curl -k -X GET -H "X-F5-Auth-Token: $token" $mgmt/shared/appsvcs/task
# delete login session
curl -k -X DELETE -H "X-F5-Auth-Token: $token" $mgmt/shared/authz/tokens/$token

**Comment by m-kratochvil** [comment]
@wncocz what happens if you don't do all these tasks periodically?

**Comment by wncocz** [comment]
> @wncocz what happens if you don't do all these tasks periodically?

By "periodically" I mean: whenever I would notice restjavad restarts on that host, or before doing a large amount of deployments.

If you read earlier in this Issue, you can see how I also had changed EXPIRE_TIME down to an hour; sometimes I even went down to 10 minutes.  This means more data would be removed from the tasks queue when I do the GET to the /mgmt.../task endpoint.

But it looks like all this can be avoided with 3.21.  I haven't done detailed analysis, and I hope I don't have to.  I'm waiting for an issue to crop up with 3.21 before I waste more time in that direction.

**Comment by m-kratochvil** [comment]
@wncocz ok, thanks for the response. I also hope no manual tasks will be required to keep the systems stable and free of junk logs etc., as I have almost 200 devices to manage..
the best would be to prevent the data-groups from generating altogether, if they only serve to keep history..
I don't like 13MB config file, especially when the config synchronizes every minute..
I'll check the situation tomorrow, I hope for improvement on 3.21..

**Comment by dstokesf5** [comment]
This has been resolved as part of the 3.21.0 release.
